### ISSUE TITLE
[EKS] [request]: Remove requirement of public IPs on EKS managed worker nodes

### ISSUE THREAD

**Issue by atheiman** [aws/containers-roadmap#607]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Tell us about your request**
Remove requirement of public IPs on EKS managed worker nodes. If worker nodes have egress access to the apiserver and the apiserver can reach the worker nodes in the same vpc by private ip, I dont think public IP should be required. Assigning public ips to k8s worker nodes (or any ec2 instances) is a security vulnerability some organizations wont accept.

**Which service(s) is this request for?**
EKS

---
*update 4/17*
We're planning to make the change to managed node groups to stop assigning public IPs to nodes on April 22, 2020.

If you are launching nodes into public subnets, you'll need to change your subnet settings to set the mapPublicIpOnLaunch to TRUE so that IPs are assigned and the nodes can connect to the public cluster endpoint. If you are not using public subnets, starting April 20, you can create a new node group and public IPs will no longer be assigned.

We'll update on this thread when the change is live.

We wrote a blog announcing this change and how to check the public IP assignment settings for your VPC.

We also wrote a deep dive blog on node networking that explains the various options for subnet and cluster endpoint configuration.

**Comment by rtripat** [comment]
We are evaluating an API opt-in to disable setting public IP on managed nodes. However, for some context on why Managed node groups get public IP in assigned in all subnets. From our [documentation](https://docs.aws.amazon.com/en_us/eks/latest/userguide/managed-node-groups.html)

"Amazon EKS managed node groups can be launched in both public and private subnets. The only requirement is for the subnets to have outbound internet access. Amazon EKS automatically associates a public IP to the instances started as part of a managed node group to ensure that these instances can successfully join a cluster.

This ensures compatibility with existing VPCs created using eksctl or the Amazon EKS-vended AWS CloudFormation templates. The public subnets in these VPCs do not have MapPublicIpOnLaunch set to true, so by default instances launched into these subnets are not assigned a public IP address. Creating a public IP address on these instances launched by managed node groups ensures that they have outbound internet access and are able to join the cluster."

**Comment by atheiman** [comment]
Yea, thats the part in your documentation I was surprised by. Seems like an unnecessary security hole going a long way to ensure nodes have outbound internet access. Public IPs on EC2s running production workloads is a security violation in lots of organizations.

I really think opting out of public IPs on worker nodes will be a requirement for many organizations to even try out managed worker nodes.

**Comment by whereisaaron** [comment]
How about defaulting to what the user asked for 😄 

If a users launches instances in a subnet they have configured with MapPublicIpOnLaunch disabled, the nodes should not get a public IP assigned by default - isn't that the whole point of the setting? It seems unhelpful and counter-intuitive to override users' express request _not_ to add public IP addresses.

Could you instead warn/inform the user in AWS console or `eksctl` output that the nodes are launching in a subnet with public IP disabled? And point them to the documentation to fix that if they made mistake by disabling public IPs?

**Comment by Siva-R** [comment]
In which version of EKS we can expect this to be fixed?

**Comment by omerfsen** [comment]
+1

**Comment by davidalger** [comment]
> We are evaluating an API opt-in to disable setting public IP on managed nodes. However, for some context on why Managed node groups get public IP in assigned in all subnets. From our documentation…

While the quoted documentation does mention `eksctl` or CF template related reasoning, but it seems rather contrived a reason. Require these networks be upgraded to utilize managed node groups and provide an upgrade path. The requirement is that "subnets have outbound internet access" which can be accomplished by having a NAT Gateway on the subnet. I.e. fulfilling the stated requirement to allow the worker node access to the control plane doesn't even need to require a Public IP be associated to the worker node.

AWS [documentation on Cluster VPC considerations](https://docs.aws.amazon.com/eks/latest/userguide/network_reqs.html) states the following:

> We recommend a network architecture that uses private subnets for your worker nodes and public subnets for Kubernetes to create internet-facing load balancers within. 

To me it seems rather bizarre that EKS best practices would outline creating both public and private subnets (with the public ones being used for things like ELB creation) yet a major managed feature addition to EKS was launched with a Public IP magically added to the EC2 instances created in a "private" subnet.

IMO an "API opt-in to disable setting public IP on managed nodes" is the wrong way to go about it. It should be opt-in to have public IP addressing applied to instances created in subnets where public addressing is turned off.

**Comment by bchav** [comment]
Hey folks, dropping in to provide some clarification around this to help with some of the security concerns here. This is indeed confusing, and we're working on product changes to make this less so, but it doesn't place your managed nodes in private subnets at additional risk. 

Due to the way VPC networking works, managed nodes with a public IP in a private subnet do not behave any differently than regular instances in private subnets. All egress to the internet would require traversal of a NAT Gateway or NAT instance, and the public IP address will not be used. Public IPs require a route to the internet via an Internet Gateway (IGW). As such, the public IP of these instances are not accessible from the internet. The public IP address is effectively a no-op. The only access to nodes in private subnets comes from other resources that share the "Cluster Security Group". 

More about the cluster security group: https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html#cluster-sg

**Comment by adonley** [comment]
Even though the above reasoning makes absolute sense, my org requires that we document and port scan any "public ip" for compliance. I guess we could write a tool that looks for nodes coming and going from node groups and listens to node group creation but it would make life easier not to deal with these issues (also it's less confusing for the security team / auditors :) )

**Comment by prashant-wipro** [comment]
Any ETA on when this fix will be made available.

**Comment by olipachi** [comment]
Like some folks have mentioned above, my org is one that explicitly blocks public IPs on EC2s. While I understand and appreciate the explanation that these public IPs are essentially no-ops, all of our security scanning tools don't feel the same way. Any ETA on when this feature would be delivered?

**Comment by abdennour** [comment]
Exposing nodes to internet is very bad. network policies can deny egress/ingress but it is at the pod level not node level.
any deadline to fix this issue ?

**Comment by zucler** [comment]
Even though nodes can be deployed into private subnets, having public IP assigned would fail security audits from any of the Fortune 500 companies. They simply wouldn't want the tools to dictate their security checklist and as right now they are strict about no public IPs.

**Comment by tabern** [comment]
Thanks everyone - your feedback is clear and we’ve been listening.

We agree its critical to respect subnet settings used for managed node groups. Our current plan is to update the behavior of managed node groups so that public IPs are not automatically assigned by managed node groups. This change will only apply to **new node groups**. Existing node groups will be unaffected.

Without taking additional steps, your EC2 instances need outbound internet access to join an EKS cluster. Nodes launched into private subnets typically access the internet through a NAT gateway, while nodes in a public subnet need an elastic IP or static public IP assigned. Some public subnets (eg: the default ones created by eksctl today) have the setting “Auto-assign Public IP Address” set to `false`. Without a public IP, these nodes will fail to become ready since they cannot access the internet. 

Once this change is made, if your nodes cannot join the cluster, want to make this simple to diagnose. We’re thinking to create a warning and health issue that explains why your nodes may not be able to connect to the cluster in the case you are launching nodes into a subnet using where “Auto-assign Public IP Address” is set to `false`. We're also planning to update our [CloudFormation VPC templates](https://docs.aws.amazon.com/eks/latest/userguide/create-public-private-vpc.html) and eksctl to set “Map Public IP address” to `true` for public subnets created using eksctl or CloudFormation. 

We are working on this with high priority and welcome your feedback.

-- Nate

**Comment by zucler** [comment]
Hi @tabern, thank you for acknowledging the issue and shedding light on the future plans.

Just want to clarify on outbound internet access requirement. Currently, it is possible to launch managed node groups into completely private subnet (no internet/NAT gateways whatsoever) by enabling private endpoint option on EKS cluster and also creating ECR and EC2 PrivateLink endpoints into the subnet and S3 gateway into the associated routing table. 

I hope that the new change won't affect this functionality and also the warning wouldn't be triggered incorrectly if nodes have no internet access, but are able to join the cluster.

**Comment by tabern** [comment]
@zucler that is correct. These are the 'additional steps' I was referencing in my last note. The change here directly enables people doing these steps today by stopping the assignment of public IPs to nodes in private subnets. However, it directly **impacts** customers running public subnets where they do not have this same setup and nodes can fail to join the cluster. Its only for these public subnets where we would show a warning. Our goal is to balance both sides here so regardless of your network setup, you get the right information.

**Comment by fcervantes2** [comment]
@tabern thank you for your explanation. Is there an ETA for when we will have access to this feature?

**Comment by cdenneen** [comment]
@abdennour 
> Exposing nodes to internet is very bad. network policies can deny egress/ingress but it is at the pod level not node level.
> any deadline to fix this issue ?

As stated by @bchav 
> All egress to the internet would require traversal of a NAT Gateway or NAT instance, and the 
> public IP address will not be used. Public IPs require a route to the internet via an Internet 
> Gateway (IGW). As such, the public IP of these instances are not accessible from the internet. 
> The public IP address is effectively a no-op.

**Comment by priyavartk** [comment]
So what I understood is that "auto assign public ip" feature has no meaning for EKS managed nodes's subnet? Its anyway going to assign them IP?

**Comment by lukemgriffith** [comment]
What is the ETA or release cadence for this product, and when can we expect this to be available. This is pretty fundamental.

**Comment by anastazya** [comment]
Error: --node-security-groups is not supported for Managed Nodegroups (--managed=true) - i'm launching the nodes in a predefined VPC that already has some servers with MySql and other services. How can i access those services from EKS if i cannot assign the existing security group ?

**Comment by rainlake** [comment]
thank you @tabern for the explaination. can you explain why "Without taking additional steps, your EC2 instances need outbound internet access to join an EKS cluster."?

I setup my control plane with both private and public access, I think worker nodes should be able to  join cluster via private access.

Have public ip assign to EC2 instances without IG seems a waste of ip address for me.

**Comment by vj03** [comment]
Any update on this issue, please? We definitely do not want an EIP for each of our worker node

**Comment by rubashanti** [comment]
+1

**Comment by scrofungulus** [comment]
Can anyone confirm that this either does or does not have an effect on how egress traffic from Nodes is routed? I'm assuming if the route table in my private subnet is directing all traffic to a NAT gateway, nothing should change then?

**Comment by web31** [comment]
+1

**Comment by cdenneen** [comment]
> Can anyone confirm that this either does or does not have an effect on how egress traffic from Nodes is routed? I'm assuming if the route table in my private subnet is directing all traffic to a NAT gateway, nothing should change then?

Correct, If “Private” network then this essentially does nothing. 

Biggest issue is removing this in order to allow privateNetworking: true

**Comment by tabern** [comment]
Hi everyone,

We're planning to make the change to managed node groups to *stop* assigning public IPs to nodes on **April 20, 2020**. 

If you are launching nodes into public subnets, you'll need to change your subnet settings to set the `mapPublicIpOnLaunch` to **TRUE** so that IPs are assigned and the nodes can connect to the public cluster endpoint. If you are not using public subnets, starting April 20, you can create a new node group and public IPs will no longer be assigned.

We'll update on this thread when the change is live.

We wrote a [blog](https://aws.amazon.com/blogs/containers/upcoming-changes-to-ip-assignment-for-eks-managed-node-groups/) announcing this change and how to check the public IP assignment settings for your VPC. 

We also wrote a [deep dive blog on node networking](https://aws.amazon.com/blogs/containers/de-mystifying-cluster-networking-for-amazon-eks-worker-nodes/) that explains the various options for subnet and cluster endpoint configuration.

- Nate

**Comment by cdenneen** [comment]
@tabern would this change allow for privateNetworking: true ?
Having to require “public” subnets in order to create managedNodeGroup is an issue that doesn’t allow the managedNodeGroup to be created from eksctl. Now clusters are partially managed via eksctl and part from console/api. 

Thanks

**Comment by tabern** [comment]
@cdenneen I'm not sure I follow. Essentially we are doing what we discussed above, respecting the IP assignment settings of the subnet. If you have public subnets and are using the cluster public endpoint, you'll need to ensure those subnets are set to assign public IPs to the nodes b/c MNG will no longer assign them. If you are using eksctl and starting nodes with private networking, you do not need to take any action - the nodes will start in private subnets and either route to the public endpoint via the subnet's nat gateway or through the private endpoint.

**Comment by cdenneen** [comment]
@tabern with eksctl it will not create a managedNodeGroup if there are no publicSubnets specified. They have made that as condition since privateNetworking couldn’t be set to true. If this is going to change then it’s something that needs to be relayed back to eksctl team for an update to coincide.

**Comment by tabern** [comment]
@cdenneen yep! We work closely with that team and this will be one of the things the change on the MNG IP assignment behavior will unblock.

**Comment by kskewes** [comment]
> We also wrote a deep dive blog on node networking that explains the various options for subnet and cluster endpoint configuration.

Can you please update the link to the deep dive blog? 404 for me.

**Comment by tabern** [comment]
Hi everyone,

We are going to delay this change until **Wednesday 4/22** in order to ensure our customers who are still creating node groups in public subnets have time to complete updating their subnet settings. One of our top priorities for EKS is operational stability, and we are making this change out of an abundance of caution in order to minimize the risk of outages for our customers.

--Nate

**Comment by bkbwese** [comment]
From the main posts 4/17 update:
> "If you are not using public subnets, starting April 20, you can create a new node group and public IPs will no longer be assigned."

@tabern I just redeployed my node groups in us-east-1 and they are still getting public IPs. Is there a specific time that this change should be live?

**Comment by lado-g** [comment]
Today i got on every eks cluster error from node-groups: AccessDenied.
All day trying to find what was the problem and found this page.
Okay, i deleted every node group and created new one in private subnet, but instances are still getting public ips, as i understand it should have only private ip, am i wrong?

**Comment by deepakpatel82** [comment]
I tried creating a new node group and i see the public ip still assigned to the worker EC2, subnets are private. Is this change in place?

**Comment by tabern** [comment]
Hey Everyone,

As of 1pm PST today, the behavior for MNG in all regions is to respect subnet settings and **not** allocate public IPs unless `mapPublicIpOnLaunch` is set to TRUE in the subnet settings.

This behavior will be in affect for all **new** managed node groups that you create on any EKS clusters. **Existing** node groups will continue to assign public IP addresses. You will need to create new node groups to get this behavior for existing clusters. 

Also, please be aware that in the case where you are attempting to create a new node group using a public subnet where `mapPublicIpOnLaunch` is set to FALSE, node group creation will fail.

#### Check the subnet setttings for your VPC:
Run: 
```
$ aws ec2 describe-subnets \
--filters "Name=vpc-id,Values=<VPC-ID>" | grep 'MapPublicIpOnLaunch\|SubnetId\|VpcId\|State'
```

#### Learn more:
* [EKS documentation](https://docs.aws.amazon.com/eks/latest/userguide/create-managed-node-group.html)
* [Blog](https://aws.amazon.com/blogs/containers/upcoming-changes-to-ip-assignment-for-eks-managed-node-groups/)
* [What's new post](https://aws.amazon.com/about-aws/whats-new/2020/04/amazon-eks-managed-node-groups-allow-fully-private-cluster-networking/)

**Comment by lado-g** [comment]
Okay, but as i wrote before i tried to create new node group in private subnet and it still gets public ip address, did this functionality release only for certain regions?

**Comment by tabern** [comment]
@lado-g this should be deployed in all regions now - we did just launch this at 1pm PST, and your comment was at 5:21AM PST, have you tried again?

**Comment by vedat227** [comment]
hey; 
I'm getting "NodeCreationFailure: Instances failed to join the kubernetes cluster" now. would this be anything todo with this change ?   private access for "API server endpoint access" is enabled

**Comment by shawnjohnson** [comment]
FWIW - I created a new "private" managed node group today, and the worker nodes came up with only private IP's - no public IP assigned - yay!

**Comment by katemarshall96** [comment]
@tabern like @vedat227  I am also getting "NodeCreationFailure: Instances failed to join the kubernetes cluster" for CloudFormation scripts that were okay working yesterday. Is there a field we can add to CloudFormation templates that will override this change to generate public IP as before?

**Comment by rtripat** [comment]
@katemarshall96 You need to set `MapPublicIpOnLaunch` to `true` for your public Subnets to generate public IP by default.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-subnet.html#cfn-ec2-subnet-mappubliciponlaunch

More details in our blog [post](https://aws.amazon.com/blogs/containers/upcoming-changes-to-ip-assignment-for-eks-managed-node-groups/).

**Comment by katemarshall96** [comment]
Thanks @rtripat - I am actually working on private subnets though

**Comment by vedat227** [comment]
> Thanks @rtripat - I am actually working on private subnets though

I'm working on private subnet as well. make sure "API server endpoint access" is enabled on the cluster and your node, in private subnet, can access to the AWS services (ecr/ecr/...) via internet gw or VPC end point. I'd double check routing tules from private subnet to NAT GW. In my case connectivity was the problem.... good luck :)

**Comment by katemarshall96** [comment]
Thanks @vedat227! I tried your suggestion with various combinations re. enabling API server endpoint access from within and outside VPC and checked routing but can't spot anything out of sorts 👎  stumped on what to do with this one...

**Comment by jodem** [comment]
@katemarshall96 check you VPC dns settings : https://docs.aws.amazon.com/eks/latest/userguide/cluster-endpoint.html

**Comment by katemarshall96** [comment]
I worked around it by using a mix of public and private subnets - couldn't get it working any more using only private subnets!

**Comment by tabern** [comment]
@katemarshall96 have you followed the steps in the 'Private Only Subnets' section of our [cluster networking blog](https://aws.amazon.com/blogs/containers/de-mystifying-cluster-networking-for-amazon-eks-worker-nodes/)? You need to ensure that ECR and EC2 private link are enabled to use only private subnets now b/c the nodes need to be able to connect to ECR and download images that we vend in order to connect to the cluster. (previously they had public IPs so this was not required).

Using public+private subnets my guess is you have a nat gateway in your public subnets which is allowing the nodes to route out to connect to ECR, etc and properly boot.

**Comment by mzs114** [comment]
@katemarshall96 
I am not sure whether this helps you now(perhaps for the new people who might face similar problem).

I was trying to setup managed nodegroups in private subnets, it kept failing(like multiple times), as suggested by one commentator, I then checked the private subnets and it had both igw and natgw in its routing tables, I removed igw and then had to try twice to get it working, it still failed the first time.

The private subnets were created(manually) and shared from another team and I had no idea this could be an issue.

Before this, I also tried enabling the private and public access to the API end point just in case, however this did not solve the problem. 
I tried it with Terraform, the `vpc_config` looks like this:

```
  vpc_config {
    endpoint_private_access = "true"
    subnet_ids              = var.all_subnet_ids
  }
```

**Comment by abdennour** [comment]
@shawnjohnson  the same as well ! I was able to assign private subnets for managed node groups.. eks 1.16 , terraform-eks module 12.1.0

**Comment by Shahard2** [comment]
What about existing nodes ?

**Comment by aktaksh** [comment]
Pub IP feature is running plenty of critical apps on worker so they should not be removed.

**Comment by vmpowercli** [comment]
Don't know if this feature has already been implemented or not but I created a new EKS cluster v1.22 with public and private subnets and the host got created in private subnets. Created an LB in public subnets and can able to hit the pod which is running on private subnets Worker nodes.
